src/smt/theory_str_noodler/formula_preprocess.cpp

```diff
@@ -1520,4 +1530,30 @@ namespace smt::noodler {
         return can_unify(left, right, check);
     }
 
+
+    /**
+     * @brief Adds restrictions from conversions to the len_formula, so that (underapproximating) unsat check can be better
+     * 
+     * Specifically, it checks if for to_code(x)/to_int(x) there is any valid word in the language of automaton for x, i.e,
+     * some one-symbol word for to_code(x) or some word containing only digits for to_int(x).
+     * 
+     * @param conversions 
+     */
+    void FormulaPreprocessor::conversions_validity(std::vector<TermConversion>& conversions) {
```

Comment:
I still wonder, whether `conversions` pass as a constructor argument. It could be more clean and it could be used in `remove_regular`.

Reply:
Yeah, I thought about that too. Also probably not contains should be part of formula.